File: elixir/test/gilded_rose_test.exs
defmodule GildedRoseTest do
  use ExUnit.Case

  import GildedRose

  @brie     "Aged Brie"
  @conjured "Conjured"
  @normal   "Normal Item"
  @sulfuras "Sulfuras, Hand of Ragnaros"
  @passes   "Backstage passes to a TAFKAL80ETC concert"

  def create_items(sell_in, quality) do
    [
      %Item{name: @sulfuras, sell_in: sell_in, quality: 80},
      %Item{name: @passes, sell_in: sell_in, quality: quality},
      %Item{name: @brie, sell_in: sell_in, quality: quality},
      %Item{name: @normal, sell_in: sell_in, quality: quality}
    ]
  end

  def elapse_days(items, 0), do: items
  
  def elapse_days(items, days) do
    elapse_days(update_quality(items), days - 1)
  end

  test "all items quality _ 11 days or more left" do
    days_left = 12
    initial_quality = 1
    result = create_items(days_left, initial_quality) |> elapse_days(1)

    assert result === [
      %Item{name: @sulfuras, sell_in: 12, quality: 80},
      %Item{name: @passes, sell_in: 11, quality: 2},
      %Item{name: @brie, sell_in: 11, quality: 2},
      %Item{name: @normal, sell_in: 11, quality: 0},
    ]
  end

  test "normal item quality _ any day greater than 0" do
    days_left = 10
    initial_quality = 10
    [_, _, _, normal] = create_items(days_left, initial_quality) |> elapse_days(1)

    assert normal === %Item{name: @normal, sell_in: 9, quality: 9}
  end

  test "normal item quality _ last day" do
    days_left = 1
    initial_quality = 5
    [_, _, _, normal] = create_items(days_left, initial_quality) |> elapse_days(1)

    assert normal === %Item{name: @normal, sell_in: 0, quality: 4}
  end
  

  test "normal item quality _ no days left" do
    days_left = 0
    initial_quality = 4
    [_, _, _, normal] = create_items(days_left, initial_quality) |> elapse_days(1)

    assert normal === %Item{name: @normal, sell_in: -1, quality: 2}
  end

  test "normal item quality _ no days left _ no more quality" do
    days_left = 0
    initial_quality = 4
    [_, _, _, normal] = create_items(days_left, initial_quality) |> elapse_days(4)

    assert normal === %Item{name: @normal, sell_in: -4, quality: 0}
  end

  test "ticket quality _ 10 days or less" do
    days_left = 10
    initial_quality = 0
    [_, ticket, _, _] = create_items(days_left, initial_quality) |> elapse_days(1)

    assert ticket === %Item{name: @passes, sell_in: 9, quality: 2}
  end

  test "ticket quality _ 5 days or less" do
    days_left = 5
    initial_quality = 0
    [_, ticket, _, _] = create_items(days_left, initial_quality) |> elapse_days(1)

    assert ticket === %Item{name: @passes, sell_in: 4, quality: 3}
  end

  test "ticket quality _ after concert" do
    days_left = 0
    initial_quality = 0
    [_, ticket, _, _] = create_items(days_left, initial_quality) |> elapse_days(1)

    assert ticket === %Item{name: @passes, sell_in: -1, quality: 0}
  end

  test "brie quality _ greater than 0 days left " do
    days_left = 1
    initial_quality = 0
    [_, _, brie, _]  = create_items(days_left, initial_quality) |> elapse_days(1)

    assert brie === %Item{name: @brie, sell_in: 0, quality: 1}
  end


  test "brie quality _ past sell in day" do
    days_left = 0
    initial_quality = 0
    [_, _, brie, _]  = create_items(days_left, initial_quality) |> elapse_days(1)

    assert brie === %Item{name: @brie, sell_in: -1, quality: 2}
  end

  test "sulfuras quality _ any days left" do
    days_left = 20
    initial_quality = 40
    [sulfuras, _, _, _]  = create_items(days_left, initial_quality) |> elapse_days(40)

    assert sulfuras === %Item{name: @sulfuras, sell_in: 20, quality: 80}
  end

  test "sulfuras quality _ lengendary quality" do
    days_left = 20
    initial_quality = 100
    [sulfuras, _, _, _]  = create_items(days_left, initial_quality) |> elapse_days(1)

    assert sulfuras === %Item{name: @sulfuras, sell_in: 20, quality: 80}
  end

  test "all items quality _ reaching max quality" do
    days_left = 5
    initial_quality = 49
    result = create_items(days_left, initial_quality) |> elapse_days(5)

    assert result === [
      %Item{name: @sulfuras, sell_in: 5, quality: 80},
      %Item{name: @passes, sell_in: 0, quality: 50},
      %Item{name: @brie, sell_in: 0, quality: 50},
      %Item{name: @normal, sell_in: 0, quality: 44}
    ]
  end


end

File: Java/src/test/java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import org.junit.jupiter.api.*;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GildedRoseTest {

    private Item[] items;
    private GildedRose app;

    @BeforeEach
    public void init() {
        items = new Item[]{
            new Item("+5 Dexterity Vest", 10, 20),
            new Item("Aged Brie", 2, 0),
            new Item("Elixir of the Mongoose", 5, 7),
            new Item("Sulfuras, Hand of Ragnaros", 0, 80),
            new Item("Sulfuras, Hand of Ragnaros", -1, 80),
            new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
            new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
            new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
            new Item("Conjured Mana Cake", 3, 6)
        };
        app = new GildedRose(items);
    }

    @Nested
    class NormalTests{
        // At the end of each day our system lowers both values (Quality and SellIn) for every item
        @Test
        public void testNormalItemQualityAndSellInDecrease() {
            app.updateQuality();
            assertEquals(19, items[0].quality, "After first day: Normal Quality should decrease by 1");
            assertEquals(9, items[0].sellIn, "After first day: Normal SellIn should decrease by 1");

            app.updateQuality();
            assertEquals(18, items[0].quality, "After second day: Normal Quality should decrease by 2");
            assertEquals(8, items[0].sellIn, "After second day: Normal SellIn should decrease by 2");
        }

        // The Quality of an item is never negative, quality of a normal item is decreasing each day
        @Test
        public void testNormalItemQualityCannotGoNegative() {
            for(int i=0; i<10; i++) {
                app.updateQuality();
            }
            assertTrue(items[2].quality > -1, "Normal Quality should never be negative");
        }

        // Once the sell by date has passed, Quality degrades twice as fast
        @Test
        public void testQualityDegradesTwiceAfterSellPassed() {
            for(int i=0; i<11; i++) {
                app.updateQuality();
            }
            assertEquals(8, items[0].quality, "Normal Quality should degrade twice after sell passed");

            app.updateQuality();
            assertEquals(6, items[0].quality, "Normal Quality should degrade twice after sell passed");
        }
    }


    @Nested
    class AgedBrieTests {
        // The Quality of an item is never more than 50, "Aged Brie" actually increases in Quality the older it gets
        @Test
        public void testAgedBrieItemQualityNeverMoreThan50() {
            for(int i=0; i<100; i++) {
                app.updateQuality();
            }
            assertTrue(items[1].quality <= 50, "Aged Brie Quality should never be more than 50");
        }

        // "Aged Brie" actually increases in Quality the older it gets
        @Test
        public void testAgedBrieItemQualityIncreases() {
            app.updateQuality();
            assertEquals(1, items[1].quality, "Aged Brie Quality should increase");

            app.updateQuality();
            assertEquals(2, items[1].quality, "Aged Brie Quality should increase");
        }
    }

    @Nested
    class SulfurasTests {
        // "Sulfuras" never has to be sold or decreases in Quality
        @Test
        @Disabled
        public void testSulfurasItemQualityUnchanged() {
        }
    }

    @Nested
    class BackstagePassesTests {
        // The Quality of an item is never more than 50, "Backstage passes" increases in Quality
        @Test
        @Disabled
        public void testBackstagePassesItemQualityNeverMoreThan50() {
        }

        // "Backstage passes" increases in Quality as its SellIn value approaches
        @Test
        @Disabled
        public void testBackstagePassesItemQualityIncreases() {
        }

        // "Backstage passes" Quality increases by 2 when there are 10 days or less
        @Test
        @Disabled
        public void testBackstagePassesItemQualityIncreasesWhenSellInLessThan11() {
        }

        // "Backstage passes" Quality increases by 3 when there are 5 days or less
        @Test
        @Disabled
        public void testBackstagePassesItemQualityIncreasesWhenSellInLessThan6() {
        }

        // "Backstage passes" Quality drops to 0 after the concert
        @Test
        @Disabled
        public void testBackstagePassesItemQualityDropsToZeroAfterConcert() {
        }
    }

    @Nested
    class ConjuredTests{
        // The Quality of an item is never negative, "Conjured" items degrade in Quality
        @Test
        @Disabled
        public void testConjuredItemQualityCannotGoNegative() {
        }

        // "Conjured" items degrade in Quality twice as fast as normal items
        @Test
        @Disabled
        public void testConjuredItemQualityDecreasesTwiceAsFast() {
        }
    }

}
